File: tutorial/sorting/quick/java/QuickSort.java
class QuickSort {
    private int partition(int[] array, int low, int high) { // A.1
        // A.2.2J
        int pivot = array[high];
        int i = low - 1;
        for(int j = low; j < high; j++) { // A.3
            if(array[j] <= pivot) {
                i++;
                swap(array, i, j);
            }
        }
        // A.4.4J
        swap(array, i+1, high);
        return i+1;
    }
    private void swap(int[] array, int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
    private void sort(int[] array, int low, int high) { // A.5
        if(low < high) { // A.6
            int part = partition(array, low, high);
            sort(array, low, part-1);
            sort(array, part+1, high);
        }
    }
    void quicksort(int[] array) {sort(array, 0, array.length-1);} // A.7
}

File: source/sort/comb/py/quicksort.py
def partition(array, low, high):  # A.1
    # A.2
    i = low - 1
    pivot = array[high]
    for j in range(low, high):  # A.3
        if array[j] <= pivot:
            i += 1
            array[i], array[j] = array[j], array[i]
    array[i + 1], array[high] = array[high], array[i + 1]  # A.4
    return i + 1


def sort(array, low, high):  # A.5
    if low < high:  # A.6
        part = partition(array, low, high)
        sort(array, low, part - 1)
        sort(array, part + 1, high)


def quicksort(array):  # A.7
    sort(array, 0, len(array) - 1)
